sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -228,7 +236,8 @@ case class ExpressionEncoder[T](
    * returns true if `T` is serialized as struct and is not `Option` type.
    */
   def isSerializedAsStructForTopLevel: Boolean = {
-    isSerializedAsStruct && !classOf[Option[_]].isAssignableFrom(clsTag.runtimeClass)
+    isSerializedAsStruct && !classOf[Option[_]].isAssignableFrom(clsTag.runtimeClass) &&
```

Comment:
`isSerializedAsStruct  && !transformerOfOption(encoder)`?

Reply:
I think it'd make sense, for the path encoder backwards compat logic I can embed / document that in shim.  The Builders could embed that.  I can take a stab at that post rc2.